NEP-641: Offchain Authorizations for Smart Contracts (#641)
near/NEPs: neps/nep-0641.md

Without a standard, each dApp must duplicate every contract's authorization rules offchain. That is fragile (contract upgrades break dApps), insecure (duplication invites divergence), and non-composable (no single library can verify a mixed delegation chain).

This NEP makes the contract the single source of truth for its account's offchain authorizations, behind one uniform view function. Resolution is recursive, so heterogeneous compositions resolve through generic caller-side logic.

### Target use cases

1. **dApp sign-in**: the dApp issues a fresh payload; the wallet contract attests it; the backend grants a session. No transaction, no access key.
2. **Offchain action approval**: authorizing an MPC signing request, a confidential swap intent, or any other offchain action.

Both reduce to one question — "does this account authorize this exact payload?" — which is what `w_resolve_auth` answers. What the payload *means* is the dApp's business, encoded in the payload itself (see [`JsonPayload`](#domain-separated-json-payload-recommended)).

## Design Decisions

<details>
<summary><strong>Verification as a view function</strong></summary>

> The governing contract is the natural authority on what authorizes its account. A view function makes that rule authoritative (no offchain library to keep in sync), composable (one call verifies any wallet type), and free (no transaction, no gas). Cost: one RPC round-trip per node of the authorization graph — one call for the common single-account case. A verifier can even skip RPC entirely by executing the contract WASM locally against fetched state (see [`fast-near`](https://github.com/vgrichina/fast-near)).

Reviewer: nit:
```suggestion
The governing contract is the natural authority on what authorizes its account. A view function makes that rule authoritative (no offchain library to keep in sync), composable (one call verifies any wallet type), and free (no transaction, no gas). Cost: one RPC round-trip per node of the authorization tree — one call for the common single-account case. A verifier can even skip RPC entirely by executing the contract WASM locally against fetched state (see [`fast-near`](https://github.com/vgrichina/fast-near)).
```

Since resolvers try [both interpretations](#resolution-of-a-single-authorization) of every blob, formats SHOULD avoid structural collisions — `AccessKeyAuthorization` rejects unknown fields and uses distinctive names for exactly this reason.

</details>

<details>
<summary><strong>Standardized signable envelope</strong></summary>

While the authorization blob is opaque, in case of Wallet contracts and Multisigs, where user signature is used to prove account ownership, adequate security measures MUST be taken to prevent replay attacks. Thus, this NEP requires the signer to include and sign the same five fields (`chain_id`, `signer_id`, `path`, `timestamp`, `payload`).

Non-signature resolvers have nothing to sign and are exempt: their authority derives from chain state alone.

</details>

<details>
<summary><strong>Caller-side recursion and the <code>path</code></strong></summary>

> A contract that cannot attest alone (multisig, extension wallet) returns its `payload` plus `pending` sub-authorizations, each with the sub-blob already **extracted** from its bundle — so the caller walks the graph without understanding any bundle format.

Reviewer: nit:
```suggestion
A contract that cannot attest alone (e.g. multisig, extension wallet) returns its `payload` plus `pending` sub-authorizations, each with the sub-blob already **extracted** from its bundle (or state) — so the caller walks the graph without understanding any bundle format.
```

</details>

<details>
<summary><strong>MPC-owned keys</strong></summary>

An account keyed by NEAR's MPC network (e.g. `v1.signer`) fits either way: the wallet contract validates the MPC-produced signature itself (MPC is just another signature scheme), or the MPC contract implements `w_resolve_auth` and is reached through normal recursion. No special handling.

</details>

## Specification

The key words "MUST", "MUST NOT", "SHOULD", "MAY" are per RFC 2119.

### Method

Implementing contracts MUST expose:

```rust
pub fn w_resolve_auth(
    path: Vec<AccountId>,
    authorization: String,
) -> AuthorizationResolution;
```

- `path` — traversal path from this contract's direct parent up to the top-level resolver; empty at top level. See [Path](#path).
- `authorization` — opaque, resolver-defined blob. See [Authorization blob](#authorization-blob).

Requirements:

- MUST be a view function: no state mutation, callable via view-call RPC.

> - MUST be deterministic over its arguments and chain state.

Reviewer: IIUC, there is no way a contract can return "non-deteministic" authorized payload if called with fixed arguments on a fixed block hash, because it has no other state to consider, no?

| `message`      | `msg.payload`                                                                                                          |
| `nonce`        | the [canonical hash](#offchainmessage) of `msg` — binds **all** envelope fields                                        |
| `recipient`    | `"<chain_id>: <signer_id>[ -> <id>]... @ <timestamp>"` — path bottom-up, timestamp truncated to whole seconds          |
| `callback_url` | `via.extra.callback_url`                                                                                               |

Example — `chain_id = "mainnet"`, `signer_id = "extension.near"`, `path = ["wallet.near", "v1.signer"]`, `timestamp = 2026-08-05T07:28:00.123456789Z`:

```text
recipient = "mainnet: extension.near -> wallet.near -> v1.signer @ 2026-08-05T07:28:00Z"
```

NEP-413 wallets render only `message` and `recipient` (`nonce` is an opaque hash), so this `recipient` puts the network, account, delegation path, and signing time in front of the user. Truncation to seconds is cosmetic; full precision stays bound through the `nonce`.

> Both `ed25519` and `secp256k1` are supported; `access_key` and `signature` curves MUST match. Existing NEP-413 wallets need no modification.

Reviewer: Can you please indicate that any digital signature algorithm can be supported this way, including ML-DSA-65?

#### Verification procedure

The resolver MUST verify an `AccessKeyAuthorization` for `account_id` at the pinned block:

1. `msg.chain_id` equals the resolver's chain ID.
2. `msg.signer_id` equals `account_id`.
3. `msg.path` equals the current resolution [path](#path).
4. The signature verifies per `via` against `access_key`.
5. `msg.timestamp` is not later than the block's timestamp.
6. `access_key` on `account_id` at the block:
   - **key exists** → accept only with `FullAccess` permission (incl. gas-key full-access); reject function-call keys;
   - **account exists, key absent** → reject — even for an implicit-derived ID, the owner may have removed the key deliberately;

>    - **account doesn't exist** → accept only if `account_id` is the implicit account derived from `access_key` (hex ed25519, or `0x` + Keccak for secp256k1) — it can be claimed under this key at any time.

Reviewer: Technically, Universal Implicit Account IDs would be derived from UniversalStateInit, so we should check if there is a known StateInit having this `access_key` that derives to `account_id`.
```suggestion
   - **account doesn't exist or not initialized yet** → accept only if `account_id` is the implicit account derived from `access_key` (hex ed25519, or `0x` + Keccak for secp256k1) — it can be claimed under this key at any time.
```

7. The result is a leaf: `{ payload: msg.payload }`.

#### Resolution of a single authorization

For every node, the resolver attempts **both** — the access-key verification above and the `w_resolve_auth` view call — optionally concurrently. A successful access-key resolution **takes precedence**. If both fail, report the access-key error when the blob parsed as `AccessKeyAuthorization` but failed key verification; otherwise the contract error.

### Pending sub-authorizations

- Empty `pending` = leaf: the contract alone authorizes `payload`.
- Non-empty `pending`: `payload` is authorized **iff** every entry, resolved recursively with the [extended path](#path), returns exactly its `expect`. Any failure or mismatch MUST fail the whole verification immediately.
- The contract MUST extract each sub-blob from its own `authorization`; the caller passes blobs on opaquely.

### Block consistency

> The resolver MUST resolve the whole graph against one block hash, SHOULD default to a recent *final* block, and SHOULD check each RPC response refers to the requested block. Historical resolution is possible (e.g. audits), but the timestamp rule keeps freshly-signed authorizations unresolvable against blocks preceding them.

Reviewer: nit:
```suggestion
The resolver MUST resolve the authorization tree against one block hash, SHOULD default to a recent *final* block, and SHOULD check each RPC response refers to the requested block. Historical resolution is possible (e.g. audits), but the timestamp rule keeps freshly-signed authorizations unresolvable against blocks preceding them.
```

        res = resolve_single(account, path, auth, block)

        if expect is not None and res.payload != expect:
            fail("payload mismatch")
        if result is None:
            result = res.payload                # top-level payload

        if res.pending is not empty:
            if len(path) >= MAX_DEPTH:          fail("max depth exceeded")
            total += len(res.pending)
            if total > MAX_SUB_AUTHORIZATIONS:  fail("too many sub-authorizations")

        sub_path = [account] + path             # prepend parent ID (bottom-up)
        for sub in res.pending:
            queue.push((sub.account_id, sub_path, sub.authorization, sub.expect))

    return result

resolve_single(account, path, auth, block):
    # both interpretations, optionally concurrent
    a = try_access_key(account, path, auth, block)
    c = try view_call(account, "w_resolve_auth",
                      { "path": path, "authorization": auth },
                      at_block = block)

    if a succeeded: return a        # full-access key takes precedence
    if c succeeded: return c
    fail(a.error if a failed access-key verification else c.error)
```

> Sibling pending entries MAY resolve concurrently. Finally, the dApp MUST check the returned payload equals the one it issued.

Reviewer: > the dApp MUST check the returned payload equals the one it issued.

Sometimes, dApp / protocol do not "issue" any payload to sign - it can be constructed on client side, and offchain verifier might only need to deserialize and use it according to its own schema (e.g. `JsonPayload` with specific `domain`, `action`, and `payload` of specific format (can itself be JSON, too)).

    Note over FE,BE: Phase 1 — Payload issuance
    FE->>BE: GET /auth/challenge
    BE->>BE: payload = JsonPayload {<br/>domain: "example.app",<br/>action: "Login",<br/>msg: "<nonce, expiry>" }<br/>store(payload, ttl=5min)
    BE-->>FE: { payload }
    end

    rect rgba(127, 127, 127, 0.15)
    Note over FE,W: Phase 2 — Authorization assembly
    FE->>W: authorize(account: "0s1234...", payload)
    W->>W: msg = OffchainMessage {<br/>chain_id: "mainnet",<br/>signer_id: "0s1234...",<br/>path: [],<br/>timestamp: now - 60s,<br/>payload }
    W->>User: render payload & confirm
    User-->>W: approve
    W->>W: proof = sign(hash(msg))
    W-->>FE: authorization =<br/>contract-defined blob(msg, proof)
    end

    rect rgba(127, 127, 127, 0.15)
    Note over FE,WC: Phase 3 — Verification
    FE->>BE: POST /auth/verify<br/>{ account_id, authorization }
    BE->>RPC: fetch latest final block
    RPC-->>BE: block_hash
    BE->>RPC: view_call("0s1234...", "w_resolve_auth",<br/>{ path: [], authorization },<br/>at_block = block_hash)
    RPC->>WC: w_resolve_auth([], authorization)
    WC->>WC: decode blob<br/>check chain_id, signer_id,<br/>path (= []), timestamp<br/>verify signature over msg

>     WC-->>RPC: { "payload": "<JsonPayload>" }

Reviewer: This renders as `{ "payload": "" }` on GitHub for some reason

    "payload": "<same payload string>"
  }
}
```

Verification:

```mermaid
sequenceDiagram
    autonumber
    participant BE as Offchain Resolver
    participant RPC as NEAR RPC<br/>(at pinned block)
    participant M as master.near
    participant E as extension.near

    Note over BE: resolve_auth(master.near, root_blob)
    BE->>RPC: fetch latest final block → block_hash
    BE->>RPC: view_call(master.near, w_resolve_auth,<br/>{ path: [], authorization: root_blob })
    RPC->>M: w_resolve_auth([], root_blob)
    M->>M: decode extension envelope<br/>check extension.near is enabled
    M-->>RPC: { payload,<br/>pending: [{ account_id: "extension.near",<br/>authorization: inner_blob,<br/>expect: payload }] }
    RPC-->>BE: { payload, pending }

    Note over BE: prepend parent: path = ["master.near"]<br/>recurse into extension.near
    BE->>RPC: view_call(extension.near, w_resolve_auth,<br/>{ path: ["master.near"],<br/>authorization: inner_blob })
    RPC->>E: w_resolve_auth(["master.near"], inner_blob)
    E->>E: check chain_id, signer_id,<br/>signed path == arg path,<br/>timestamp not in future<br/>verify signature
    E-->>RPC: { payload }
    RPC-->>BE: { payload }

>     Note over BE: payload == expect ✓<br/>no more pending → done<br/>dApp checks payload == issued payload

Reviewer: This note can look a bit misleading as it sounds like dApp checks the payload returned from the `extension.near`, not from `master.near`

## Drawbacks

- **Recursion complexity**: callers orchestrate recursive view calls instead of one check — encapsulated in a library, but heavier than NEP-413.
- **Per-contract blob formats**: signing clients need format-specific encoders per wallet type. Unavoidable given wallet heterogeneity.
- **Latency**: nested delegation costs one RPC round-trip per level (siblings parallelize; parent/child cannot). Depth rarely exceeds 2.
- **Borsh + SHA3 in clients**: the canonical hash requires both in every signing client, including non-NEAR-native ones.
- **Uninitialized deterministic accounts**: a NEP-616 account existing only as a `StateInit` derivation cannot answer view calls, so its contract authorizations cannot resolve yet (access-key authorizations for implicit accounts are exempt). Needs RPC support for view calls against pre-initialized state.

## Future Possibilities

- **Pre-init view calls**: RPC execution against a supplied `StateInit` (NEP-616), enabling resolution for not-yet-deployed deterministic accounts.
- **More `AccessKeySchema` variants**, including post-quantum (e.g. ML-DSA).

> - **Local WASM execution** of `w_resolve_auth` in verifier SDKs (e.g. [`fast-near`](https://github.com/vgrichina/fast-near)) — no per-node RPC.

Reviewer: Note that WASM bytecode currently deployed to the account is actually a part of its on-chain state and cannot be "cached" unconditionally.

- **Composable**: any caller can verify any account uniformly, regardless of wallet type.
- **Cheap**: view calls do not require transactions, gas payment, or chain-state changes.

The trade-off is that resolution requires an RPC round-trip per node in the authorization graph. For typical authentication flows (single account, single call) this is one RPC. For deeply nested multisigs it scales linearly with graph size, which is acceptable for off-chain flows.

</details>

<details>
<summary><strong>Opaque authorization blob</strong></summary>

Each wallet contract defines its own internal format for the authorization blob: an EIP-712 envelope for an Ethereum-signing wallet, a WebAuthn assertion for a passkey wallet, a Borsh-encoded bundle of member authorizations for a multisig, and so on. This NEP intentionally does not standardize the blob format.

Standardizing across all wallet types would require either lowest-common-denominator constraints (limiting innovation) or a fat schema (unmanageable). Companion NEPs may standardize specific blob formats for common schemes (EIP-712, WebAuthn, etc.) without modifying this NEP.

</details>

<details>

> <summary><strong>Two purposes: <code>PROVE_OWNERSHIP</code> and <code>APPROVE_OFFCHAIN_ACTION</code></strong></summary>

Author: I have already received a piece of feedback that the purpose might be unnecessary:

My original idea was to allow multisig members to sign in as DAO and for that one won't need to collect M-of-N signatures from other members, while for other actions you definitely want to require M-of-N consensus. For that, dApp should be able to signal to the wallet which purpose it is, and then the w_auth_resolve function to validate the provided authorization against the purpose.

Thus, the multisig contract could have defined two additional kinds of actions: "prove_ownership" and "approve_offchain_action", and then verify the policy requirements (1-of-N or M-of-N, etc) in the w_auth_resolve function.

However, multisig like SputnikDAO has quite a few kinds of actions (change policy, create proposal, approve proposal, etc), and it seems that it could be quite limiting to have only two kinds of actions. Still, the "function call" proposal is also quite generic, and does not allow bringing further granularity in the policy enforcement.

That said, for wallet contracts, where there is a single signer, PROVE_OWNERSHIP and APPROVE_OFFCHAIN_ACTION are redundant, and the purpose field only brings semantic difference that could be communicated to the user.

I still believe there is value in the purpose field even if it feels redundant in some cases and limiting on others.

There is a new emerging concept for SputnikDAO - automated members (steward contracts, that cast their vote on ping if the proposal matches their internal auto-approval policy). For those cases, w_auth_resolve("PROVE_OWNERSHIP") would always return INVALID, while for w_auth_resolve("APPROVE_OFFCHAIN_ACTION") it would depend on the action - if the authorization would have the action that is in line with the auto-approval policy, then the contract would return RESOLVED.

- **Multisig contracts** that require M-of-N approval from member accounts, each of which may itself be a wallet contract of a different type.
- **Future schemes** — ZK-proof-gated accounts, attestation-gated accounts, session-key delegated accounts — implemented entirely in contract code.

Today, dApps integrating with such accounts must hand-roll per-contract verification logic off-chain, duplicating the contract's authorization rules. This is fragile (contract upgrades break dApps), insecure (off-chain duplication invites divergence), and does not compose (a multisig of mixed wallet types cannot be verified by any single library).

This NEP makes the **on-chain contract the single source of truth** for whether an off-chain authorization is valid for its account, accessible via a uniform view-function interface. Resolution is recursive, so heterogeneous compositions resolve correctly through generic caller-side logic.

## Design Considerations

The following considerations shaped the design. Each is documented so future revisions and implementers can see the trade-offs.

<details>

> <summary><strong>Verification as a view function on the contract</strong></summary>

Author: The client-side SDK for verification could even have an implementation that would run the wasm contract locally on the client side, especially for well-known wallet contracts.

Here is a Node.js implementation for view-function calls: https://github.com/vgrichina/fast-near
